src/main/java/io/jenkins/plugins/httpclient/RobustHTTPClient.java

```diff
@@ -0,0 +1,258 @@
+                                            diag = IOUtils.toString(err, contentEncoding != null ? contentEncoding.getValue() : null);
+                                        }
+                                    } else {
+                                        diag = null;
+                                    }
+                                    throw new AbortException(String.format("Failed to %s, response: %d %s, body: %s", whatVerbose, responseCode.get(), statusLine != null ? statusLine.getReasonPhrase() : "?", diag));
+                                }
+                                connectionUser.use(response);
+                            }
+                        }
+                        return null; // success
+                    }).get(timeout, TimeUnit.SECONDS);
+                } catch (TimeoutException x) {
+                    throw new ExecutionException(new IOException(x)); // ExecutionException unwrapped & treated as retryable below
+                }
+                listener.getLogger().flush(); // seems we can get interleaved output with master otherwise
+                return; // success
+            } catch (ExecutionException wrapped) {
+                Throwable x = wrapped.getCause();
+                if (x instanceof IOException) {
+                    if (attempt == stopAfterAttemptNumber) {
+                        throw (IOException) x; // last chance
+                    }
+                    if (responseCode.get() > 0 && responseCode.get() < 200 || responseCode.get() >= 300 && responseCode.get() < 500) {
+                        throw (IOException) x; // 4xx errors should not be retried
+                    }
+                    // TODO exponent base could be made into a configurable parameter
+                    Thread.sleep(Math.min(((long) Math.pow(2d, attempt)) * waitMultiplier, waitMaximum * 1000));
+                    listener.getLogger().printf("Retrying %s after: %s%n", whatConcise, x instanceof AbortException ? x.getMessage() : x.toString());
+                    attempt++; // and continue
+                } else if (x instanceof InterruptedException) { // all other exceptions considered fatal
+                    throw (InterruptedException) x;
+                } else if (x instanceof RuntimeException) {
+                    throw (RuntimeException) x;
+                } else if (x != null) {
+                    throw new RuntimeException(x);
+                } else {
+                    throw new IllegalStateException();
+                }
+            }
+        }
+    }
+
+    /**
+     * Mask out query string details in a URL.
+     * Useful in conjunction with {@code VirtualFile#toExternalURL}.
+     * @param url any URL
+     * @return the same, but with any query string concealed
+     */
+    public static String sanitize(URL url) {
+        return url.toString().replaceFirst("[?].+$", "?…");
```

Comment:
not sure, but Could be interesting to remove the stuff between @ and //? I mean, if we have the URL `http://U:P@HOST?QUERY` we replace it with `http://U:P@HOST?...`, it is possible to make the same with the user and password and transform the URL to  `http://..@HOST?...`

Reply:
passwords should be removed, user would be ok 

also, should it be three chars `...` or one unicode char `…`? I think the former

Reply:
`URL.authority` is generally opaque, not necessarily `user:pass`. It could be masked if needed. Currently there is no use case for doing so, as (semi-)secrets are passed in the query string.